File: patch_resolution.go
package main

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

const (
	exeName      = "HUNDRED_LINE.exe"
	backupName   = "HUNDRED_LINE.exe.backup"
	configName   = "userconfig.properties"
	configBackup = "userconfig.properties.backup"

	targetWidth  = 5120
	targetHeight = 2880

	origWidth  = 3840
	origHeight = 2160

	// Offsets in the executable
	tableOffset        = 0xBDA4F0
	widthGetterOffset  = 0x054DF7
	widthParamOffset   = 0x4B4305
	heightGetterOffset = 0x054D37
	heightParamOffset  = 0x4B430B
)

type Patch struct {
	Offset      int
	Original    []byte
	Replacement []byte
	Description string
}

func main() {
	fmt.Println("╔════════════════════════════════════════════════════════════╗")
	fmt.Println("║     The Hundred Line - Resolution Patcher                  ║")
	fmt.Println("║     Patches 3840x2160 → 5120x2880 (5K)                     ║")
	fmt.Println("╚════════════════════════════════════════════════════════════╝")
	fmt.Println()

	// Check if exe exists
	if _, err := os.Stat(exeName); os.IsNotExist(err) {
		fmt.Printf("ERROR: %s not found!\n", exeName)
		fmt.Println("Please place this patcher in the same folder as the game executable.")
		waitForExit()
		return
	}

	// Read the executable to check if patched
	data, err := os.ReadFile(exeName)
	if err != nil {
		fmt.Printf("ERROR: Failed to read %s: %v\n", exeName, err)
		waitForExit()
		return
	}

	if isPatched(data) {
		runRestoreMode()
	} else {
		runPatchMode(data)
	}
}

func isPatched(data []byte) bool {
	newWidthBytes := make([]byte, 4)
	binary.LittleEndian.PutUint32(newWidthBytes, targetWidth)

	// Check if the resolution table has the patched value
	if len(data) > tableOffset+4 {
		return bytes.Equal(data[tableOffset:tableOffset+4], newWidthBytes)
	}
	return false
}

func runPatchMode(data []byte) {
	fmt.Println("Mode: PATCH")
	fmt.Println()
	fmt.Printf("This will patch %s to support 5120x2880 resolution.\n", exeName)
	fmt.Println("A backup will be created automatically.")
	fmt.Println()

	fmt.Printf("File size: %d bytes\n", len(data))
	fmt.Println()

	// Create patches
	patches := createPatches(data)
	if len(patches) == 0 {
		fmt.Println("ERROR: No valid patch locations found.")
		fmt.Println("The executable may be a different version.")
		waitForExit()
		return
	}

	// Display patches
	fmt.Println("The following changes will be made:")
	fmt.Println("────────────────────────────────────────────────────────────")
	for _, p := range patches {
		origVal := binary.LittleEndian.Uint32(p.Original)
		newVal := binary.LittleEndian.Uint32(p.Replacement)
		fmt.Printf("  0x%08X: %d → %d\n", p.Offset, origVal, newVal)
		fmt.Printf("              %s\n", p.Description)
	}
	fmt.Println("────────────────────────────────────────────────────────────")
	fmt.Printf("Total patches: %d\n", len(patches))
	fmt.Println()
	fmt.Println("userconfig.properties will also be updated.")
	fmt.Println()

	// Wait for confirmation
	fmt.Println("Press ENTER to apply the patch, or close this window to cancel...")
	waitForEnter()

	// Create backup
	fmt.Printf("Creating backup: %s\n", backupName)
	if err := os.WriteFile(backupName, data, 0644); err != nil {
		fmt.Printf("ERROR: Failed to create backup: %v\n", err)
		waitForExit()
		return
	}

	// Apply patches
	patchedData := make([]byte, len(data))
	copy(patchedData, data)

	for _, p := range patches {
		copy(patchedData[p.Offset:], p.Replacement)
	}

	// Write patched executable
	if err := os.WriteFile(exeName, patchedData, 0644); err != nil {
		fmt.Printf("ERROR: Failed to write patched file: %v\n", err)
		waitForExit()
		return
	}

	// Update userconfig.properties
	if err := patchUserConfig(); err != nil {
		fmt.Printf("WARNING: Failed to update %s: %v\n", configName, err)
		fmt.Println("You may need to manually set the resolution in the config file.")
	} else {
		fmt.Printf("Updated %s\n", configName)
	}

	fmt.Println()
	fmt.Println("╔════════════════════════════════════════════════════════════╗")
	fmt.Println("║                    PATCH SUCCESSFUL!                       ║")
	fmt.Println("╚════════════════════════════════════════════════════════════╝")
	fmt.Println()
	fmt.Println("Run this patcher again to restore the original executable.")
	fmt.Println()
	waitForExit()
}

func runRestoreMode() {
	fmt.Println("Mode: RESTORE")
	fmt.Println()
	fmt.Println("The executable appears to be patched.")
	fmt.Println("This will restore the original unpatched executable.")
	fmt.Println()

	// Check if backup exists
	if _, err := os.Stat(backupName); os.IsNotExist(err) {
		fmt.Printf("ERROR: Backup file not found: %s\n", backupName)
		fmt.Println("Cannot restore without backup.")
		waitForExit()
		return
	}

	// Verify backup is readable
	backupData, err := os.ReadFile(backupName)
	if err != nil {
		fmt.Printf("ERROR: Failed to read backup: %v\n", err)
		waitForExit()
		return
	}

	fmt.Printf("Backup size: %d bytes\n", len(backupData))
	fmt.Println()

	// Wait for confirmation
	fmt.Println("Press ENTER to restore the original executable, or close this window to cancel...")
	waitForEnter()

	// Restore from backup
	if err := os.WriteFile(exeName, backupData, 0644); err != nil {
		fmt.Printf("ERROR: Failed to restore executable: %v\n", err)
		waitForExit()
		return
	}

	// Remove exe backup
	if err := os.Remove(backupName); err != nil {
		fmt.Printf("WARNING: Failed to remove backup file: %v\n", err)
	}

	// Restore userconfig.properties if backup exists
	if err := restoreUserConfig(); err != nil {
		fmt.Printf("WARNING: Failed to restore %s: %v\n", configName, err)
	} else {
		fmt.Printf("Restored %s\n", configName)
	}

	fmt.Println()
	fmt.Println("╔════════════════════════════════════════════════════════════╗")
	fmt.Println("║                   RESTORE SUCCESSFUL!                      ║")
	fmt.Println("╚════════════════════════════════════════════════════════════╝")
	fmt.Println()
	fmt.Println("The original executable has been restored.")
	fmt.Println()
	waitForExit()
}

func patchUserConfig() error {
	// Read existing config
	configData, err := os.ReadFile(configName)
	if err != nil {
		return fmt.Errorf("failed to read config: %w", err)
	}

	// Parse JSON
	var config map[string]interface{}
	if err := json.Unmarshal(configData, &config); err != nil {
		return fmt.Errorf("failed to parse config: %w", err)
	}

	// Backup original config (only if backup doesn't exist)
	if _, err := os.Stat(configBackup); os.IsNotExist(err) {
		if err := os.WriteFile(configBackup, configData, 0644); err != nil {
			return fmt.Errorf("failed to create config backup: %w", err)
		}
		fmt.Printf("Created config backup: %s\n", configBackup)
	}

	// Update values (preserving X and Y)
	config["App.Window.W"] = targetWidth
	config["App.Window.H"] = targetHeight
	config["App.Window.Mode"] = "BorderlessWindowed"

	// Write updated config with indentation
	updatedData, err := json.MarshalIndent(config, "", "\t")
	if err != nil {
		return fmt.Errorf("failed to encode config: %w", err)
	}

	if err := os.WriteFile(configName, updatedData, 0644); err != nil {
		return fmt.Errorf("failed to write config: %w", err)
	}

	return nil
}

func restoreUserConfig() error {
	// Check if config backup exists
	if _, err := os.Stat(configBackup); os.IsNotExist(err) {
		return nil // No backup to restore, not an error
	}

	// Read backup
	backupData, err := os.ReadFile(configBackup)
	if err != nil {
		return fmt.Errorf("failed to read config backup: %w", err)
	}

	// Restore config
	if err := os.WriteFile(configName, backupData, 0644); err != nil {
		return fmt.Errorf("failed to restore config: %w", err)
	}

	// Remove config backup
	if err := os.Remove(configBackup); err != nil {
		fmt.Printf("WARNING: Failed to remove config backup: %v\n", err)
	}

	return nil
}

func createPatches(data []byte) []Patch {
	var patches []Patch

	origWidthBytes := make([]byte, 4)
	origHeightBytes := make([]byte, 4)
	newWidthBytes := make([]byte, 4)
	newHeightBytes := make([]byte, 4)

	binary.LittleEndian.PutUint32(origWidthBytes, origWidth)
	binary.LittleEndian.PutUint32(origHeightBytes, origHeight)
	binary.LittleEndian.PutUint32(newWidthBytes, targetWidth)
	binary.LittleEndian.PutUint32(newHeightBytes, targetHeight)

	// Resolution table patches
	if len(data) > tableOffset+8 {
		if bytes.Equal(data[tableOffset:tableOffset+4], origWidthBytes) {
			patches = append(patches, Patch{
				Offset:      tableOffset,
				Original:    origWidthBytes,
				Replacement: newWidthBytes,
				Description: fmt.Sprintf("Resolution table: %d → %d (width)", origWidth, targetWidth),
			})
		}
		if bytes.Equal(data[tableOffset+4:tableOffset+8], origHeightBytes) {
			patches = append(patches, Patch{
				Offset:      tableOffset + 4,
				Original:    origHeightBytes,
				Replacement: newHeightBytes,
				Description: fmt.Sprintf("Resolution table: %d → %d (height)", origHeight, targetHeight),
			})
		}
	}

	// Code patches
	codePatches := []struct {
		offset      int
		original    []byte
		replacement []byte
		description string
	}{
		{widthGetterOffset, origWidthBytes, newWidthBytes, "Width getter (mov eax, 3840)"},
		{widthParamOffset, origWidthBytes, newWidthBytes, "Resolution param width (mov edx, 3840)"},
		{heightGetterOffset, origHeightBytes, newHeightBytes, "Height getter (mov eax, 2160)"},
		{heightParamOffset, origHeightBytes, newHeightBytes, "Resolution param height (mov r8d, 2160)"},
	}

	for _, cp := range codePatches {
		if len(data) > cp.offset+4 && bytes.Equal(data[cp.offset:cp.offset+4], cp.original) {
			patches = append(patches, Patch{
				Offset:      cp.offset,
				Original:    cp.original,
				Replacement: cp.replacement,
				Description: cp.description,
			})
		} else {
			fmt.Printf("WARNING: Mismatch at 0x%X, skipping: %s\n", cp.offset, cp.description)
		}
	}

	return patches
}

func waitForEnter() {
	reader := bufio.NewReader(os.Stdin)
	reader.ReadString('\n')
}

func waitForExit() {
	fmt.Println("Press ENTER to exit...")
	waitForEnter()
}

func init() {
	// Change to the directory where the executable is located
	exe, err := os.Executable()
	if err == nil {
		dir := filepath.Dir(exe)
		os.Chdir(dir)
	}
}

File: patch_resolution.py
#!/usr/bin/env python3
"""
Resolution Patch for The Hundred Line -Last Defense Academy-

This script patches the game executable to support higher resolutions
like 5120x2880 (5K) by modifying hardcoded resolution values.

BACKUP YOUR EXECUTABLE BEFORE RUNNING THIS!

Usage:
    python patch_resolution.py HUNDRED_LINE.exe [--dry-run]
"""

import struct
import argparse
import shutil
from pathlib import Path
from dataclasses import dataclass


@dataclass
class Patch:
    offset: int
    original: bytes
    replacement: bytes
    description: str


def create_patches(data: bytes, target_width: int, target_height: int) -> list[Patch]:
    """Create patches for resolution modification"""
    
    patches = []
    
    # Patch only the 3840x2160 entry in the resolution table
    print(f"\nPatching 3840x2160 -> {target_width}x{target_height}")
    
    orig_width = struct.pack('<I', 3840)
    orig_height = struct.pack('<I', 2160)
    new_width = struct.pack('<I', target_width)
    new_height = struct.pack('<I', target_height)
    
    # Resolution table entry
    table_offset = 0xBDA4F0
    if data[table_offset:table_offset + 4] == orig_width:
        patches.append(Patch(table_offset, orig_width, new_width, 
            f"Resolution table: 3840 -> {target_width} (width)"))
    if data[table_offset + 4:table_offset + 8] == orig_height:
        patches.append(Patch(table_offset + 4, orig_height, new_height,
            f"Resolution table: 2160 -> {target_height} (height)"))
    
    # Code patches for resolution limits
    code_patches = [
        (0x054DF7, orig_width, new_width, "Width getter (mov eax, 3840)"),
        (0x4B4305, orig_width, new_width, "Resolution param width (mov edx, 3840)"),
        (0x054D37, orig_height, new_height, "Height getter (mov eax, 2160)"),
        (0x4B430B, orig_height, new_height, "Resolution param height (mov r8d, 2160)"),
    ]
    
    for offset, orig, new, desc in code_patches:
        if data[offset:offset + 4] == orig:
            patches.append(Patch(offset, orig, new, desc))
        else:
            print(f"WARNING: Mismatch at 0x{offset:X}, skipping: {desc}")
    
    return patches


def apply_patches(data: bytearray, patches: list[Patch], dry_run: bool = False) -> bytearray:
    """Apply patches to the executable data"""
    
    print("\n" + "=" * 60)
    print("PATCHES TO APPLY:" if not dry_run else "PATCHES (DRY RUN):")
    print("=" * 60)
    
    for patch in patches:
        orig_val = struct.unpack('<I', patch.original)[0]
        new_val = struct.unpack('<I', patch.replacement)[0]
        print(f"  0x{patch.offset:08X}: {orig_val} -> {new_val}")
        print(f"                   {patch.description}")
        
        if not dry_run:
            data[patch.offset:patch.offset + len(patch.replacement)] = patch.replacement
    
    print(f"\nTotal patches: {len(patches)}")
    return data


def main():
    parser = argparse.ArgumentParser(
        description="Patch The Hundred Line for higher resolution support",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
    python patch_resolution.py HUNDRED_LINE.exe --dry-run
    python patch_resolution.py HUNDRED_LINE.exe
    python patch_resolution.py HUNDRED_LINE.exe --width 7680 --height 4320
        """
    )
    parser.add_argument("exe_path", type=Path, help="Path to HUNDRED_LINE.exe")
    parser.add_argument("--dry-run", action="store_true", help="Analyze only, don't modify")
    parser.add_argument("--width", type=int, default=5120, help="Target width (default: 5120)")
    parser.add_argument("--height", type=int, default=2880, help="Target height (default: 2880)")
    parser.add_argument("--no-backup", action="store_true", help="Skip creating backup")
    
    args = parser.parse_args()
    
    if not args.exe_path.exists():
        print(f"Error: File not found: {args.exe_path}")
        return 1
    
    print(f"Resolution Patch for The Hundred Line")
    print(f"Target resolution: {args.width}x{args.height}")
    print("=" * 60)
    
    with open(args.exe_path, 'rb') as f:
        data = bytearray(f.read())
    
    print(f"File size: {len(data):,} bytes")
    
    patches = create_patches(bytes(data), args.width, args.height)
    
    if not patches:
        print("\nNo patches could be created. The executable may be different from expected.")
        return 1
    
    apply_patches(data, patches, args.dry_run)
    
    if args.dry_run:
        print("\n[DRY RUN] No changes made.")
        return 0
    
    if not args.no_backup:
        backup_path = args.exe_path.with_suffix('.exe.backup')
        if not backup_path.exists():
            print(f"\nCreating backup: {backup_path}")
            shutil.copy2(args.exe_path, backup_path)
        else:
            print(f"\nBackup already exists: {backup_path}")
    
    with open(args.exe_path, 'wb') as f:
        f.write(data)
    
    print("\n" + "=" * 60)
    print("DONE!")
    print("=" * 60)
    print(f"\nUpdate your userconfig.properties:")
    print(f'  "App.Window.W" : {args.width},')
    print(f'  "App.Window.H" : {args.height},')
    print(f"\nRestore from backup if needed:")
    print(f"  copy {args.exe_path.with_suffix('.exe.backup')} {args.exe_path}")
    
    return 0


if __name__ == "__main__":
    exit(main())